src/timepicker/TimePicker.ts

```diff
@@ -0,0 +1,359 @@
+		};
+	});
+}
+
+/**
+ * @private
+ * Create a numeric timestamp for the specified hour, minute, and second units.
+ *
+ * @param units   An object containing the hours, minutes, and seconds for the time.
+ * @param date    An optional date object.
+ * @return        The timestamp, in milliseconds, according to universal time.
+ */
+function getTime(units: TimeUnits, date = new Date()) {
+	const { hour, minute, second } = units;
+	date.setHours(hour);
+	date.setMinutes(minute as number);
+	date.setSeconds(second as number);
+	date.setMilliseconds(0);
+	return date.getTime();
+}
+
+/**
+ * Convert a standard time string into an object with `hour`, `minute`, and `second` number properties.
+ *
+ * For example, '12:30' is converted to `{ hour: 12, minute: 30, second: 0 }`, and '19:03:27' is converted
+ * to `{ hour: 19, minute: 3, second: 27 }`.
+ *
+ * @param value   A standard time string or an object with `hour`, `minute`, and `second` properties.
+ * @throws Error  For any string not in the format 'HH:mm' or 'HH:mm:ss'.
+ * @return        An object containing `hour`, `second`, and `number` properties.
+ */
+export function parseUnits (value: string | TimeUnits): TimeUnits {
+	if (typeof value === 'string') {
+		if (!TIME_PATTERN.test(value)) {
+			throw new Error('Time strings must be in the format HH:mm or HH:mm:ss');
+		}
+
+		const [ hour, minute, second = 0 ] = value.split(':').map(unit => parseInt(unit, 10));
+		return { hour, minute, second } as TimeUnits;
+	}
+
+	return value;
+}
+
+export const TimePickerBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export class TimePicker extends TimePickerBase<TimePickerProperties> {
+	private _getOptions: () => TimeUnits[];
+	private _nativeInputNode: HTMLInputElement;
+
+	render(): DNode {
+		const {
+			disabled,
+			formId,
+			invalid,
+			label,
+			readOnly,
+			required,
+			useNativeElement
+		} = this.properties;
+
+		if (useNativeElement) {
+			const input = this.renderNativeInput();
+			let children: DNode[] = [ input ];
+
+			if (label) {
+				children = [ w(Label, {
+					bind: this,
+					classes: this.classes(
+						disabled ? css.disabled : null,
+						invalid ? css.invalid : null,
+						readOnly ? css.readonly : null,
+						required ? css.required : null
+					),
+					formId,
+					label
+				}, [ input ]) ];
+			}
+
+			return v('span', {
+				classes: this.classes(css.root),
+				key: 'root'
+			}, children);
+		}
+
+		return this.renderCustomInput();
+	}
+
+	protected onElementCreated(element: HTMLElement, key: string) {
+		if (key === 'native-input') {
+			this._nativeInputNode = element as HTMLInputElement;
+		}
+	}
+
+	@onPropertiesChanged()
+	protected onPropertiesChanged(event: PropertiesChangeEvent<this, TimePickerProperties>) {
+		if (
+			!this._getOptions ||
+			includes(event.changedPropertyKeys, 'start') ||
+			includes(event.changedPropertyKeys, 'end') ||
+			includes(event.changedPropertyKeys, 'step')) {
+			this._getOptions = (() => {
```

Comment:
`_getOptions` caches the list of options after creating it so it doesn't have to generate the default 1,440 values each time the results menu is shown. If either `start`, `end`, or `step` changes, then that cached list needs to be regenerated.

Alternatively, I could instead add a private `_options` instance variable so that it's the list that gets invalidated rather than the callback.